lib/internal/mime.js

```diff
@@ -80,43 +87,82 @@ function parseTypeAndSubtype(str) {
   };
 }
 
-const EQUALS_SEMICOLON_OR_END = /[;=]|$/;
-const QUOTED_VALUE_PATTERN = /^(?:([\\]$)|[\\][\s\S]|[^"])*(?:(")|$)/u;
-
-function removeBackslashes(str) {
-  let ret = '';
-  // We stop at str.length - 1 because we want to look ahead one character.
-  let i;
-  for (i = 0; i < str.length - 1; i++) {
-    const c = str[i];
-    if (c === '\\') {
-      i++;
-      ret += str[i];
+function isSemicolonOrEquals(char) {
+  const code = StringPrototypeCharCodeAt(char, 0)
+  return code === SEMICOLON_CHARCODE || code === EQUALS_CHARCODE;
+}
+
+function isQuoteOrBackslash(char) {
+  const code = StringPrototypeCharCodeAt(char, 0)
+  return code === QUOTE_CHARCODE || code === BACKSLASH_CHARCODE;
+}
+
+function isNotWhiteSpace(char) {
+  return !RegExpPrototypeTest(WHITESPACE_REGEX, char);
+}
+
+function moveUntil(str, predicate, marker, {
+  toASCII,
+  removeBackslashes,
+} = {
+  toASCII: false,
+  removeBackslashes: false,
+},
+) {
+  const start = marker.position;
+  const length = str.length;
+  let index = start;
+  let result = '';
+  let char = str[index];
+
+  while (predicate(char) === false && index < length) {
+    let asCharCode;
+    let controlCharCode;
+    // We lazily assign it if needed, as not all the iterations
+    // might require it
+    if (removeBackslashes && (
+        asCharCode = StringPrototypeCharCodeAt(char, 0)
+        ) === BACKSLASH_CHARCODE
+    ) {
+      // Jump to the next char
+      index++;
+      char = str[index];
+      controlCharCode = StringPrototypeCharCodeAt(char, 0);
+    }
+
+    if (toASCII) {
+      asCharCode = controlCharCode != null ? controlCharCode : asCharCode || StringPrototypeCharCodeAt(char, 0)
+      result += asCharCode >= ASCII_LOWER_LOWERBOUND &&
+        asCharCode <= ASCII_LOWER_UPPERBOUND ?
+        StringPrototypeToLowerCase(char) :
+        char;
     } else {
-      ret += c;
+      result += char;
     }
+
+    index++;
+    char = str[index];
   }
-  // We add the last character if we didn't skip to it.
-  if (i === str.length - 1) {
-    ret += str[i];
-  }
-  return ret;
-}
 
+  marker.position = index;
+
+  return result;
+}
 
 function escapeQuoteOrSolidus(str) {
   let result = '';
   for (let i = 0; i < str.length; i++) {
     const char = str[i];
-    result += (char === '"' || char === '\\') ? `\\${char}` : char;
+    const code = StringPrototypeCharCodeAt(char, 0)
```

Comment:
Nit: Ideally, if we don't have quote or backslash character, we don't need to create a result string and append it for each character. I don't know the probability of having this happy path, but if it's common, we might just check on the beginning and return the `str`

Reply:
I had in mind using the `Regex#matchAll` but it requires having the Regex marked as global (making them stateful), which can make usages of Regex tricky. Matter of tradeoffs maybe?

Though, this only affects while using the `toString` method of the class (template literals, concatenation, etc.). At parsing is not involved at all 🙂

Reply:
`RexExp.prototype.matchAll` creates a `RegExpStringIterator` object, and it has all kinds of prototype pollution vulnerabilities. I don't think it's the right tool for `lib/` (and doubt it would be more performant than the alternatives).